### PR TITLE
Add --nodns=strict for #1750 (environments with proxy and no public DNS services)

### PR DESCRIPTION
**Pull request by mzpqnxow**
This is one way to implement a fix for #1750

It's certainly not elegant, though the code changes are minimal

It requires regression testing and would benefit from feedback on a better way to approach it. It's very difficult working on such a gigantic shell script :)

EDIT: I'm open to suggestions for better ways to implement it. Like I said, it's very hard to trace the flow so I took the path of least resistance as I saw it

### CODE REVIEW

**Comment by drwetter** [comment]
ACK & got it. It takes a while until I am able to look into it. 

One thing which was obvious is the code style. We use double square brackets and inside those if no quoting is required unless you need to. Otherwise inside double brackets every single \c\h\a\r\ \i\s \q\u\o\t\e\d during runtime.

**Comment by mzpqnxow** [comment]
> ACK & got it. It takes a while until I am able to look into it.

No rush, we're all busy

> One thing which was obvious is the code style. We use double square brackets and inside those if no quoting is required unless you need to. Otherwise inside double brackets every single \c\h\a\r\ \i\s \q\u\o\t\e\d during runtime.

Gotcha, I'll update it to be consistent with that. I'm usually pretty strict about adhering to project style, sorry 'bout that

**Comment by mzpqnxow** [comment]
I fixed those issues and also removed 2 lines of whitespace that got added somehow. Would you like me to add something to the documentation or is it OK to worry about that later? I'm not sure anyone else is going to be using this feature anyway since I'm the only one to have ever complained about the problem :)

File: testssl.sh [hunk 1]
```diff
@@ -20374,8 +20380,14 @@ determine_ip_addresses() {
      local ip4=""
      local ip6=""
 
+     if [[ "$NODNS" == "strict" ]]; then
```
Comment (drwetter) [inline testssl.sh:20383]: here and some instances below the quotes aren't necessary

**Comment by drwetter** [comment]
> Would you like me to add something to the documentation or is it OK to worry about that later

good point. One step at a time

File: testssl.sh (hunk 1 above)
Comment (mzpqnxow) [reply testssl.sh:20383]: I just pushed a commit to clean all of these up

**Review by mzpqnxow — COMMENTED** [review]
All set with those

**Comment by drwetter** [comment]
Sorry for the delay.

First of all: basically ``--n  none`` should have sufficed. If not I would suggest to identify the places where this isn't enough. Then decide whether we want to improve those occurrences or introduce another flag.

Then: as indicated you put lines like ``if [[ "$NODNS" == strict ]]`` in low level functions like ``determine_optimal_proto`` or ``run_server_preference()`` rather than using the calling function(s). 

(Minor thing: Indentation is 5 chars, no 2 YAML style here)

So as indicated overall it doesn't make much sense to me, thus I rather close this.

**Comment by mzpqnxow** [comment]
@drwetter I'm confused- you're saying you don't want to accept it? Either I'm misunderstanding or possibly you're not remembering the issue that started this whole thing- unfortunately, it's a long discussion - https://github.com/drwetter/testssl.sh/issues/1750. We discussed at length:

1. The issue
2. The potential fixes

The summary of the issue: When openssl.sh is run in a very restrictive environment (e.g. enterprise environment with no access to public DNS resolvers, only HTTP proxy only for access to the Internet) it bombs out even when you direct it to inhibit DNS. It makes at least one (for what reasons I now don't remember, but it was not necessary)

There were a few hacks to work around the issue- some things like setting the IP address to 0.0.0.0 or something similar, or adding a dummy entry to /etc/hosts every time to you want to assess an endpoint

If you wouldn't mind, can you just confirm that you glanced through that issue so we're on the same page?

Thanks!

**Comment by drwetter** [comment]
See comments above. It's hackish.

Please also see here https://github.com/drwetter/testssl.sh/issues/1750#issuecomment-998817440 .

Don't know we're on the same page but I respect and understood you were saying. :-). Not sure whether you got my points though? The last paragraph, see link above, provides a solution though.

**Comment by mzpqnxow** [comment]
> First of all: basically ``--n  none`` should have sufficed

Yes, I felt the same way :)

> If not I would suggest to identify the places where this isn't enough.
> Then decide whether we want to improve those occurrences or introduce another flag.

From the linked issue (your comments)

> TL;DR:
> testssl.sh --proxy=http://<your_proxyip> -n none --ip=0.0.0.0 https://www.homedepot.com works but requires a confirmation:
 > Your OpenSSL cannot connect to 0.0.0.0:443
 > The results might look ok but they could be nonsense. Really proceed ? ("yes" to continue) --> yes
> wheres ./testssl.sh --proxy=http://<your_proxyip> -n none https://www.homedepot.com doesn't work.

> ./testssl.sh --proxy=auto -n none --ip=0.0.0.0 https://www.homedepot.com works when http_proxy is set to an IP.

> It should work also when in /etc/hosts DNS resolution is supplied.

I had pointed those out as well and asked "do you want a PR, intend to "fix" it yourself or do you want to leave the workarounds as the solution or say it's won't fix/as expected," to which you said "there will be a proper solution, and I'll see what you propose," hence this PR (admittedly 9 months or so late, I forgot to commit the changes back to my repo after they worked in my environment)

I need this either way because of my use-case, which is calling it from a restricted environment from a web application, where HTTP proxies have DNS names. None of the workarounds we discussed in the original issue were great for me

- Hard-coding the proxy IP causes spurious failures because the HTTP proxies in my environment are referenced by a DNS name, to facilitate round-robin and maintenance. I could add application logic in my application to resolve the address immediately before calling testssl.sh, though

- It's not practical for me to add entries to /etc/hosts dynamically. It's just a bad idea. I also don't can't get the IP of the site from my environment

- Having to keep a pipe open to send a newline to the prompt is clunky (though you can say my PR is clunky too, I was aiming for a "proper" solution rather than a workaround)

> So as indicated overall it doesn't make much sense to me, thus I rather close this.

Fair enough, sorry for the confusion and probably the formatting (I'm on mobile)

Thanks for the project and for your patience

**Comment by mzpqnxow** [comment]
> I would suggest to identify the places where this isn't enough. Then decide whether we want to improve those occurrences or introduce another flag.

FWIW- that's exactly what this is/was, though in retrospect making it a third option for --nodns was probably not the right choice and might have been the cause for requiring so many changes throughout the flow

**Comment by drwetter** [comment]
hi @mzpqnxow ,

thanks for your reply.

As said technically your PR works with the constraint that it only tests one hostname using random IPs. That's a constraint of using a proxy resolving hostnames, not your PR.

I understand that it is not easy for you finding the right spot where to add the lines to make it work. 

#1750 is still open and if not you or anybody else can';t do this, I'll find the time probably.

Thanks!
